### ISSUE TITLE
intermittent test_signalKILL failure on BSD

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4186]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4186|
|Type|defect|
|Created|2009-12-23 20:34:43Z|
|Branch|https://github.com/twisted/twisted/tree/test-signalkill-4186|

This failure shows up not infrequently on at least the FreeBSD builders:

```
===============================================================================
[ERROR]: twisted.test.test_process.PosixProcessTestCase.test_signalKILL

Traceback (most recent call last):
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/process.py", line 240, in doRead
    return fdesc.readFromFD(self.fd, self.dataReceived)
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/fdesc.py", line 94, in readFromFD
    callback(output)
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/process.py", line 243, in dataReceived
    self.proc.childDataReceived(self.name, data)
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/process.py", line 745, in childDataReceived
    self.proto.childDataReceived(name, data)
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/protocol.py", line 451, in childDataReceived
    self.outReceived(data)
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/test/test_process.py", line 274, in outReceived
    self.transport.signalProcess(self.signal)
  File "/usr/home/wiqd/twisted/freebsd-py2.6-select/Twisted/twisted/internet/process.py", line 322, in signalProcess
    raise ProcessExitedAlready()
twisted.internet.error.ProcessExitedAlready: 
-------------------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__4186 4186
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_test_signalkill_4186 branches/test-signalkill-4186
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1261600483000000 1261600483000000
changetime__1261926392000000 1261926392000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27863]) Branching to 'test-signalkill-4186'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27865]) Keep track of whether the child has been signaled yet and don't signal it again if we get more output

refs [#4186](https://github.com/twisted/twisted/issues/4186)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

I instrumented the test to report when output is received and when the process ended event arrives.  Right away I saw that sometimes the "\n" from the child is delivered separately from the rest of the line.  This would cause `signalProcess` to be called twice, which isn't what's intended, and could lead to this error, I think.  A few builds on a !FreeBSD builder after changing the test to only try to deliver the signal once passed.  It might be good to have someone with BSD access run this a few more times to be sure.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix set owner to @exarkun|
|-|-|

Well, this change seems reasonable to me, since it's really just making some code obviously more robust in the face of fragmented data.

1. You modified some methods which didn't have docstrings and didn't add docstrings to them.

That's all I have. +1 if you add those docstrings.

I haven't actually run these tests on any BSD machines, though. They pass on my machine.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [27890]) Add a couple docstrings to SignalProtocol

refs [#4186](https://github.com/twisted/twisted/issues/4186)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [27892]) Merge test-signalkill-4186

Author: exarkun
Reviewer: radix
Fixes: [#4186](https://github.com/twisted/twisted/issues/4186)

Make several of the tests for child process functionality more robust by
having them account for the possibility that a write from the child they
interact with might be delivered in multiple chunks to the process
protocol.

This should fix a handful of intermittent `ProcessAlreadyExited` errors
in tests on `PosixProcessTestCase`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
